### PR TITLE
General cleanup

### PR DESCRIPTION
**Pull request by ayushkarnawat**
## What does this PR do?
- [x] Update `setup.py` to include installation requirements based on the python version
- [x] Adds docstrings for all (necessary) methods/classes
- [x] Add type hints for all functions (python 3.x). If user is on python version 2.x, then the package `typing` is (now) required (see Line9-10 in `setup.py`) to use this package.

### CODE REVIEW

**Comment by ayushkarnawat** [comment]
There are a couple of things that I was unsure about. For example should we have separate install requirements for devs, i.e. `pylint`, `pytorch`, `crc32c`, `pytest`, `pytest-cov`, `pytest-pep8`, since they will all be required to run unittest cases (after we design them).

Additionally, potentially out of scope for this PR, but once we have the unittest cases, should we use a CI program to run all test to verify that (1) a predefined code-style is followed, and (2) all the test cases pass for this build. Please let me know if this is over optimzing too early?

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,107 @@
...
+.ropeproject
+
+# mkdocs documentation
+/site
+
+# mypy
+.mypy_cache/
+
+# VScode workspace settings
+.vscode/
```
Comment (vahidk) [inline .gitignore]: Please revert these.

File: tfrecord/iterator_utils.py [hunk 2]
```diff
@@ -1,37 +1,74 @@
-import typing
+"""Iterator utils."""
+
+import warnings
+from typing import Any, Callable, Iterable, Iterator, List
```
Comment (vahidk) [inline tfrecord/iterator_utils.py]: import typing.
I prefer not to pollute the namespace.
Do this for the other places as well, import modules not types/classes/objects.

File: .gitignore (hunk 1 above)
Comment (ayushkarnawat) [reply .gitignore]: Just for clarification, do you mean all the changes in the .gitignore file, or just the workspace specific ones?

Comment (vahidk) [reply .gitignore]: I meant all. I don't think it's necessary.

Comment (ayushkarnawat) [reply .gitignore]: Hmm, how about including some of the python packaging things (i.e. `*.egg-info/`)? These get created when installing the package locally through `python setup.py [install|develop]`

File: tfrecord/iterator_utils.py [hunk 3]
```diff
@@ -1,37 +1,76 @@
...
-def sample_iterators(iterators: typing.List[typing.Iterable], ratios):
-    """Create an iterator by sampling from a list of iterators with given ratios."""
-    iterators = [cycle(iterator) for iterator in iterators]
+    def cycle(iterator_fn: typing.Callable):
+        """Create a repeating iterator from an iterator generator."""
+        while True:
+            for element in iterator_fn():
+                yield element
+
+    itrs = [cycle(iterator) for iterator in itrs]
```
Comment (vahidk) [inline tfrecord/iterator_utils.py]: Revert itrs to iterators.

File: tfrecord/iterator_utils.py [hunk 4]
```diff
@@ -1,37 +1,76 @@
...
+    -------
+    item: Any
+        Decoded bytes of the features into its respective data type (for
+        an individual record) from an iterator.
+    """
     buffer = []
     try:
         for _ in range(queue_size):
-            buffer.append(next(iterator))
+            buffer.append(next(itr))
```
Comment (vahidk) [inline tfrecord/iterator_utils.py]: Revert itr to iterator.

File: tfrecord/iterator_utils.py [hunk 5]
```diff
@@ -1,37 +1,76 @@
...
+    item: Any
+        Decoded bytes of features into its respective data types from
+        an iterator (based off their sampling ratio).
+    """
 
-def sample_iterators(iterators: typing.List[typing.Iterable], ratios):
-    """Create an iterator by sampling from a list of iterators with given ratios."""
-    iterators = [cycle(iterator) for iterator in iterators]
+    def cycle(iterator_fn: typing.Callable):
+        """Create a repeating iterator from an iterator generator."""
```
Comment (vahidk) [inline tfrecord/iterator_utils.py]: Move back cycle to module scope.

**Comment by vahidk** [comment]
I merged your changes and after testing locally found errors in the code. So I had to revert. Please create the pull requests again and make sure you test your changes.

**Pull request merged**